File: frontend/src/types/order.ts
export type ShippingMethod = 'pickup' | 'delivery'
export type PaymentMethod = 'cash' | 'paypal'
export type PaymentStatus = 'pending' | 'paid' | 'failed'
export type OrderStatus = 'initiated' | 'sent_to_whatsapp' | 'confirmed' | 'shipped' | 'delivered' | 'cancelled'

export interface OrderCustomer {
  name: string
  email?: string
  phone: string
  address?: string
  custom_fields: Record<string, unknown>
}

export interface OrderItem {
  product_id: string
  name: string
  size?: string
  color?: string
  quantity: number
  unit_price: number
  line_total: number
}

export interface Order {
  id: string
  store_id: string
  order_number: string
  customer: OrderCustomer
  items: OrderItem[]
  currency: string
  subtotal: number
  shipping_method: ShippingMethod
  shipping_fee: number
  discount_amount: number
  coupon_code?: string
  total: number
  payment_method: PaymentMethod
  payment_status: PaymentStatus
  status: OrderStatus
  track_token: string
  whatsapp_url?: string
  created_at: string
  updated_at: string
}

export interface OrderItemCreate {
  product_id: string
  size?: string
  color?: string
  quantity: number
}

export interface OrderCustomerCreate {
  name: string
  email?: string
  phone: string
  address?: string
  custom_fields?: Record<string, unknown>
}

export interface OrderCreate {
  items: OrderItemCreate[]
  customer: OrderCustomerCreate
  shipping_method: ShippingMethod
  coupon_code?: string
  payment_method?: PaymentMethod
}

export interface OrderUpdate {
  status?: OrderStatus
  payment_status?: PaymentStatus
}

export interface OrderTracking {
  order_number: string
  status: OrderStatus
  items: OrderItem[]
  total: number
  currency: string
  created_at: string
  store_name: string
  store_whatsapp: string
}

File: backend/app/models/order.py
from pydantic import BaseModel, Field
from typing import Optional, List, Dict, Any
from datetime import datetime
from enum import Enum
from bson import ObjectId


class PyObjectId(ObjectId):
    @classmethod
    def __get_validators__(cls):
        yield cls.validate

    @classmethod
    def validate(cls, v, handler):
        if not ObjectId.is_valid(v):
            raise ValueError("Invalid ObjectId")
        return ObjectId(v)

    @classmethod
    def __get_pydantic_json_schema__(cls, core_schema, handler):
        return {"type": "string"}


class ShippingMethodEnum(str, Enum):
    PICKUP = "pickup"
    DELIVERY = "delivery"


class PaymentMethodEnum(str, Enum):
    CASH = "cash"
    PAYPAL = "paypal"


class PaymentStatusEnum(str, Enum):
    PENDING = "pending"
    PAID = "paid"
    FAILED = "failed"


class OrderStatusEnum(str, Enum):
    INITIATED = "initiated"
    SENT_TO_WHATSAPP = "sent_to_whatsapp"
    CONFIRMED = "confirmed"
    SHIPPED = "shipped"
    DELIVERED = "delivered"
    CANCELLED = "cancelled"


class OrderCustomer(BaseModel):
    name: str
    email: Optional[str] = None
    phone: str
    address: Optional[str] = None
    custom_fields: Dict[str, Any] = {}


class OrderItem(BaseModel):
    product_id: PyObjectId
    name: str
    size: Optional[str] = None
    color: Optional[str] = None
    quantity: int
    unit_price: float
    line_total: float


class OrderModel(BaseModel):
    """Order model."""

    id: Optional[PyObjectId] = Field(default=None, alias="_id")
    store_id: PyObjectId
    order_number: str

    customer: OrderCustomer
    items: List[OrderItem]

    currency: str = "INR"
    subtotal: float
    shipping_method: ShippingMethodEnum
    shipping_fee: float = 0.0
    discount_amount: float = 0.0
    coupon_code: Optional[str] = None
    total: float

    payment_method: PaymentMethodEnum = PaymentMethodEnum.CASH
    payment_status: PaymentStatusEnum = PaymentStatusEnum.PENDING
    paypal_order_id: Optional[str] = None

    status: OrderStatusEnum = OrderStatusEnum.INITIATED
    track_token: str

    whatsapp_message: Optional[str] = None

    created_at: datetime = Field(default_factory=datetime.utcnow)
    updated_at: datetime = Field(default_factory=datetime.utcnow)

    class Config:
        populate_by_name = True
        arbitrary_types_allowed = True
        json_encoders = {ObjectId: str}
